cdap-app-fabric/src/main/java/co/cask/cdap/security/impersonation/DefaultUGIProvider.java

```diff
@@ -199,51 +241,160 @@ private File localizeKeytab(Location keytabLocation) throws IOException {
+    
+    if (entityId instanceof TwillRunProgramId) {
+      programRunId = getProgramRunIdFromTwillRunProgramId(entityId);
+      if (programRunId == null) {
+        return Collections.emptyMap();
+      }
+    } else {
+      programRunId = ((ProgramRunId) entityId);
+    }
           
-          if (runRecord != null) {
-              Gson gson = new Gson();
-              Type stringStringMap = new TypeToken<Map<String, String>>() { }.getType();
-
-              Map<String, String> properties = runRecord.getProperties();
-              String runtimeArgsJson = properties.get("runtimeArgs");
-              if (runtimeArgsJson != null) {
-                  properties = gson.fromJson(runtimeArgsJson, stringStringMap);
-                  if ((properties.containsKey(RUNTIME_ARG_KEYTAB)) && 
-                          (properties.containsKey(RUNTIME_ARG_PRINCIPAL))) {
-                      LOG.debug("Found impersonation info from runtime arguments");
-                      return properties;
-                  } else {
-                      LOG.debug("Could not find keytab, principal in runtime args");
-                  }
-              } else {
-                  LOG.debug("Could not find any runtime args");
-              }
-          } else {
-              LOG.debug("Could not obtain program's meta run record");
-          }
-      } else {
-          LOG.debug("Entity Id not of type ProgramRunId, skippinh checking of runtime args");
+    RunRecordMeta runRecord = null;
+    long sleepDelayMs = TimeUnit.MILLISECONDS.toMillis(100);
+    long timeoutMs = TimeUnit.SECONDS.toMillis(10);
+
+    try {
+      runRecord = Retries.callWithRetries(() -> {
+        RunRecordMeta rec = store.getRun(programRunId);
+        if (rec != null) {
+          return rec;
+        }
+        throw new Exception("Retrying again to fetch run record");
+      }, RetryStrategies.timeLimit(timeoutMs, TimeUnit.MILLISECONDS,
+                  RetryStrategies.fixDelay(sleepDelayMs, TimeUnit.MILLISECONDS)),
+                  Exception.class::isInstance);
+    } catch (Exception e) {
+      LOG.warn("Failed to fetch run record for {} due to {}", programRunId, e.getMessage(), e);
+      return Collections.emptyMap();
+    }
+
+    Map<String, String> resultProps = Collections.emptyMap();
+    Type stringStringMap = new TypeToken<Map<String, String>>() { }.getType();
+    Map<String, String> properties = runRecord.getProperties();
+    LOG.debug("Runtime args for program {} are sysargs {}, properties {}", programRunId,
+            Arrays.toString(runRecord.getSystemArgs().entrySet().toArray()),
+            Arrays.toString(properties.entrySet().toArray()));
+    String runtimeArgsJson = properties.get("runtimeArgs");
+    if (runtimeArgsJson != null) {
+      properties = gson.fromJson(runtimeArgsJson, stringStringMap);
+      if (properties.containsKey(ProgramOptionConstants.LOGGED_IN_USER)) {
+        properties.remove(ProgramOptionConstants.LOGGED_IN_USER);
```

Comment:
why are we removing it?

Reply:
To avoid any conflicts with user supplied runtime arguments, if same named property is provided by user as user runtime arguments, remove it from our map. I am adding the same as comment in code.